src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -99,13 +107,24 @@ ucs_status_t uct_cuda_copy_ep_put_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov, si
 {
 
     uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_copy_iface_t);
+    uct_cuda_copy_stream_desc_t *cuda_stream;
     ucs_status_t status;
 
+    /* get cuda stream object */
+    cuda_stream = ucs_mpool_get(&iface->cuda_stream_h2d_desc);
+    if (ucs_unlikely(cuda_stream == NULL)) {
+        ucs_error("Failed to allocate cuda stream object");
+        return UCS_ERR_NO_MEMORY;
+    }
+
     status = uct_cuda_copy_post_cuda_async_copy(tl_ep, (void *)remote_addr,  iov[0].buffer,
                                                 iov[0].length, cudaMemcpyHostToDevice,
-                                                iface->stream_h2d,
+                                                cuda_stream->stream,
                                                 &iface->outstanding_h2d_cuda_event_q, comp);
 
+    /* release stream object */
+    ucs_mpool_put(cuda_stream);
```

Comment:
can we really release it immediately and not wait for a completion? so how it is different than previous situation?

Reply:
yes yossie.  Before, streams objects  are created in iface_init(), so even through test (MTT) is pure CPU application,  all 48 ranks sharing the GPU and  creating the resources and which caused the GPU out of memory.  This PR creates stream only on first use which never happen for these MTT runs.

In real GPU/MPI uses cases,   usually each MPI ranks exclusively uses the GPU and offloads the kernel to it.  GPU context sharing with multiple ranks is not usual case.

Reply:
so why need an mpool? seems only one is used. can keep it on iface, but just create it on demand

Reply:
I think creating/destroying each time is costly operation. otherwise, we need to add a if condition check first time and create it  instead of mpoo_get().

Reply:
yes, need to add if condition. but ucs_mpool_get() also has if condition to check if the pool is not empty.

Reply:
choose  mpool also to play with more streams in future.   But, I think for now will create on demand and will go to mpool in future when need to come. will update PR

Reply:
fixed.